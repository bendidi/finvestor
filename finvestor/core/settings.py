import logging
from datetime import timedelta
from typing import Literal, Optional

from pydantic import AnyHttpUrl, AnyUrl, BaseSettings, validator

logger = logging.getLogger(__name__)


class FinvestorConfig(BaseSettings):
    # DATABASE SETTINGS
    FINVESTOR_DATABASE_URI: Optional[AnyUrl]

    ###############################################
    #     EXTARNAL DATA PROVIDERS SETTINGS        #
    ###############################################

    # YAHOO FINANCE BASE URL TO GET PRICE DATA
    YF_BASE_URL: AnyHttpUrl = AnyHttpUrl(
        url="https://query2.finance.yahoo.com",
        scheme="https",
        host="query2.finance.yahoo.com",
    )
    # YAHOO FINANCE SCRAPE URL TO GET TICKER INFO AND FUNDAMENTALS
    YF_SCRAPE_URL: AnyHttpUrl = AnyHttpUrl(
        url="https://finance.yahoo.com/quote",
        scheme="https",
        host="finance.yahoo.com/quote",
    )
    # BUSINESS INSIDER URL TO SCRAP ISIN OF TICKERS
    ISIN_SCRAPE_URL: AnyHttpUrl = AnyHttpUrl(
        url="https://markets.businessinsider.com/ajax/SearchController_Suggest",
        scheme="https",
        host="markets.businessinsider.com",
        path="/ajax/SearchController_Suggest",
    )

    # ALPACA settings to connect to their api
    APCA_API_BASE_URL: AnyHttpUrl = AnyHttpUrl(
        url="https://paper-api.alpaca.markets",
        scheme="https",
        host="paper-api.alpaca.markets",
    )
    APCA_API_DATA_URL: AnyHttpUrl = AnyHttpUrl(
        url="https://data.alpaca.markets",
        scheme="https",
        host="data.alpaca.markets",
    )
    # by defult free tier has 15 minue delay
    APCA_HISTORICAL_DATA_DELAY_SECONDS: timedelta = timedelta(seconds=15 * 60)
    APCA_API_KEY_ID: Optional[str] = None
    APCA_API_SECRET_KEY: Optional[str] = None

    DATA_PROVIDER: Literal[None, "alpaca", "yahoo_finance"] = None

    @validator("DATA_PROVIDER", pre=True)
    def data_provider_setter(cls, data_provider, values):
        key_id = values.get("APCA_API_KEY_ID")
        secret_id = values.get("APCA_API_SECRET_KEY")
        if data_provider is None:
            if key_id is not None and secret_id is not None:
                _provider = "alpaca"
            else:
                _provider = "yahoo_finance"
            logger.debug(f"Using '{_provider}' as market data provider.")
            return _provider

        if data_provider == "alpaca" and (key_id is None or secret_id is None):
            raise ValueError(
                "To use 'alpaca' as market data provider, you need to set up: "
                "'APCA_API_KEY_ID' and 'APCA_API_KEY_ID'"
            )
        return data_provider

    class Config:
        case_sensitive = True
        env_file = ".env"
        env_file_encoding = "utf-8"


config = FinvestorConfig()
