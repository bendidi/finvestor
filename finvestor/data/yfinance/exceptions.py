class EmptyResponse(Exception):
    pass


class ErrorResponse(Exception):
    pass
