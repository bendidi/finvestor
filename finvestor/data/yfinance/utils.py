import logging
from datetime import datetime, timedelta
from typing import Dict, List, Literal, Union

import pytz

from finvestor.models import ValidInterval

logger = logging.getLogger(__name__)

ValidRange = Literal[
    "1d", "5d", "1mo", "3mo", "6mo", "1y", "2y", "5y", "10y", "ytd", "max"
]


INTERVAL_TO_MAX_DAYS: Dict[ValidInterval, float] = {
    "1m": 7,
    "2m": 59.9,
    "5m": 59.9,
    "15m": 59.9,
    "30m": 59.9,
    "1h": 729.9,
}


class _ValidIntervals:
    _max_days_to_valid_intervals: Dict[float, List[ValidInterval]] = {
        7: ["1m", "2m", "5m", "15m", "30m", "1h", "1d", "5d", "1mo", "3mo"],
        59.9: ["2m", "5m", "15m", "30m", "1h", "1d", "5d", "1mo", "3mo"],
        729.9: ["1h", "1d", "5d", "1mo", "3mo"],
    }
    _default: List[ValidInterval] = ["1d", "5d", "1mo", "3mo"]

    @classmethod
    def from_range(cls, range: ValidRange) -> List[ValidInterval]:
        delta = time_symbol_to_timedelta(range)
        for max_days, valid_intervals in cls._max_days_to_valid_intervals.items():
            if delta.days <= max_days:
                return valid_intervals
        return cls._default

    @classmethod
    def from_datetime(cls, _datetime: datetime) -> List[ValidInterval]:
        now = pytz.utc.localize(datetime.utcnow())
        delta = now - _datetime
        for max_days, valid_intervals in cls._max_days_to_valid_intervals.items():
            if delta.days <= max_days:
                return valid_intervals
        return cls._default


def time_symbol_to_timedelta(symbol: Union[ValidInterval, ValidRange]) -> timedelta:
    if symbol.endswith("m"):
        return timedelta(minutes=float(symbol[:-1]))
    elif symbol.endswith("h"):
        return timedelta(hours=1)
    elif symbol.endswith("d"):
        return timedelta(days=float(symbol[:-1]))
    elif symbol.endswith("mo"):
        return timedelta(days=float(symbol[:-2]) * 30.417)
    elif symbol.endswith("y"):
        return timedelta(days=float(symbol[:-1]) * 365)
    elif symbol == "ytd":
        return timedelta(days=365)
    elif symbol == "max":
        return timedelta(days=365 * 10)

    raise ValueError(f"Unsupported symbol: {symbol}")


def get_valid_intervals_for_range(
    interval: ValidInterval, range: ValidRange
) -> List[ValidInterval]:
    valid_intervals = _ValidIntervals.from_range(range)
    if interval in valid_intervals:
        idx = valid_intervals.index(interval)
        return valid_intervals[idx:]
    return valid_intervals


def get_valid_intervals_for_datetime(
    interval: ValidInterval, _datetime: datetime
) -> List[ValidInterval]:
    valid_intervals = _ValidIntervals.from_datetime(_datetime)
    if interval in valid_intervals:
        idx = valid_intervals.index(interval)
        return valid_intervals[idx:]
    return valid_intervals


def coerce_end_time_to_interval(
    ticker: str, interval: ValidInterval, start_date: datetime, end_date: datetime
) -> datetime:

    delta = time_symbol_to_timedelta(interval)

    if start_date + delta < end_date:
        return end_date

    logger.debug(
        f"[{ticker}]: Interval={interval} is too big for provided start={start_date} "  # type: ignore # noqa
        f"-> end={end_date}, updating end_date to {start_date + delta}"
    )
    return start_date + delta
