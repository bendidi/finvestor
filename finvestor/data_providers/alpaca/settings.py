from datetime import timedelta

from pydantic import AnyHttpUrl, BaseSettings


class APCASettings(BaseSettings):

    BASE_URL: AnyHttpUrl = AnyHttpUrl.build(
        scheme="https",
        host="paper-api.alpaca.markets",
    )
    DATA_URL: AnyHttpUrl = AnyHttpUrl.build(
        scheme="https",
        host="data.alpaca.markets",
    )
    # by defult free tier has 15 minue delay
    # If provided as an ENV var, it should be seconds
    # ex: export ALPACA_HISTORICAL_DATA_DELAY=900
    HISTORICAL_DATA_DELAY: timedelta = timedelta(0)
    API_KEY_ID: str
    API_SECRET_KEY: str

    class Config:
        env_prefix = "ALPACA_"
        case_sensitive = True
        env_file = ".env"
        env_file_encoding = "utf-8"


alpaca_settings = APCASettings()

if __name__ == "__main__":
    print(alpaca_settings)
