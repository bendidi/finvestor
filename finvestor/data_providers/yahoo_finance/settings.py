from pydantic import AnyHttpUrl, BaseSettings


class YFSettings(BaseSettings):
    BASE_URL: AnyHttpUrl = AnyHttpUrl.build(
        scheme="https",
        host="query2.finance.yahoo.com",
    )

    # YAHOO FINANCE SCRAPE URL TO GET TICKER INFO AND FUNDAMENTALS
    SCRAPE_URL: AnyHttpUrl = AnyHttpUrl.build(
        scheme="https",
        host="finance.yahoo.com/quote",
    )
    # BUSINESS INSIDER URL TO SCRAP ISIN OF TICKERS
    ISIN_SCRAPE_URL: AnyHttpUrl = AnyHttpUrl.build(
        scheme="https",
        host="markets.businessinsider.com",
        path="/ajax/SearchController_Suggest",
    )

    class Config:
        env_prefix = "YAHOO_FINANCE_"
        case_sensitive = True
        env_file = ".env"
        env_file_encoding = "utf-8"


yf_settings = YFSettings()

if __name__ == "__main__":
    print(yf_settings)
