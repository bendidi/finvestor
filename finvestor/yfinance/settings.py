from pydantic import AnyHttpUrl, BaseSettings

YF_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36"  # noqa
}


class YFSettings(BaseSettings):
    YAHOO_FINANCE_BASE_URL: AnyHttpUrl = AnyHttpUrl(
        url="https://query2.finance.yahoo.com",
        scheme="https",
        host="query2.finance.yahoo.com",
    )

    YAHOO_FINANCE_SCRAPE_URL: AnyHttpUrl = AnyHttpUrl(
        url="https://finance.yahoo.com/quote",
        scheme="https",
        host="finance.yahoo.com/quote",
    )
    ISIN_SCRAPE_URL: AnyHttpUrl = AnyHttpUrl(
        url="https://markets.businessinsider.com/ajax/SearchController_Suggest",
        scheme="https",
        host="markets.businessinsider.com",
        path="/ajax/SearchController_Suggest",
    )
    ISIN_SCRAPE_MAX_RESULTS: int = 25
    ISIN_SCRAPE_TIMEOUT: int = 5

    class Config:
        case_sensitive = True
