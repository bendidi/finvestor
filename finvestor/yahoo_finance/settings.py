from pydantic import AnyHttpUrl

from finvestor.core import BaseConfig


class YFConfig(BaseConfig, env_prefix="YAHOO_FINANCE_"):
    BASE_URL: AnyHttpUrl = AnyHttpUrl.build(
        scheme="https",
        host="query2.finance.yahoo.com",
    )

    # YAHOO FINANCE SCRAPE URL TO GET TICKER INFO AND FUNDAMENTALS
    SCRAPE_URL: AnyHttpUrl = AnyHttpUrl.build(
        scheme="https",
        host="finance.yahoo.com/quote",
    )
    # BUSINESS INSIDER URL TO SCRAP ISIN OF TICKERS
    ISIN_SCRAPE_URL: AnyHttpUrl = AnyHttpUrl.build(
        scheme="https",
        host="markets.businessinsider.com",
        path="/ajax/SearchController_Suggest",
    )


config = YFConfig()
