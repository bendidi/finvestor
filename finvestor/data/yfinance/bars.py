import asyncio
import logging
from datetime import datetime
from typing import Any, Dict, List, Optional, Tuple

import numpy as np
import pytz
from httpx import AsyncClient
from pydantic import validate_arguments

from finvestor.core import config
from finvestor.data.yfinance import YF_HEADERS
from finvestor.data.yfinance.exceptions import EmptyResponse, ErrorResponse
from finvestor.data.yfinance.utils import (
    ValidInterval,
    ValidRange,
    coerce_end_time_to_interval,
    get_valid_intervals_for_datetime,
    get_valid_intervals_for_range,
    time_symbol_to_timedelta,
)
from finvestor.models import Bar, Bars

logger = logging.getLogger(__name__)


async def get_quotes(
    ticker: str,
    *,
    client: AsyncClient,
    start: Optional[datetime],
    end: Optional[datetime],
    range: Optional[ValidRange],
    interval: ValidInterval = "1m",
    include_prepost: bool = False,
    coerce_interval: bool = False,
) -> Tuple[List[float], Dict]:

    params: Dict[str, Any] = {}

    if start is not None:
        assert start.tzinfo == pytz.utc, (
            f"Only timezone aware datetimes with tz={pytz.UTC} are supported, "
            f"got: start<{start.tzinfo}>"
        )
        if end is not None:
            assert end.tzinfo == pytz.utc, (
                f"Only timezone aware datetimes with tz={pytz.UTC} are supported, "
                f"got: end<{end.tzinfo}>"
            )
            assert end > start, (
                f"Provided end datetime ({end}) should be larger than start "
                f"datetime ({start})"
            )
            if coerce_interval:
                end = coerce_end_time_to_interval(ticker, interval, start, end)
        else:
            end = start + time_symbol_to_timedelta(interval)
        params["period1"] = int(datetime.timestamp(start))
        params["period2"] = int(datetime.timestamp(end))
    else:
        params["range"] = range

    params["interval"] = interval
    params["includePrePost"] = include_prepost
    params["events"] = "div,splits"
    resp = await client.get(
        url=f"{config.YF_BASE_URL}/v8/finance/chart/{ticker}",
        params=params,
        headers=YF_HEADERS,
    )
    response = resp.json()

    error = response.get("chart", {}).get("error")
    if error:
        logger.error(f"Error response from Y-Finance [{ticker}]: {error}")
        raise ErrorResponse(error)

    result = response.get("chart", {}).get("result", [])
    if not len(result):
        logger.error(
            f"Empty response from Y-Finance [{ticker}]: range={range} | "
            f"interval={interval} | start={start} | end={end}"
        )
        raise EmptyResponse(f"Empty response: {response}")

    timestamps = result[0].get("timestamp")
    if not timestamps:
        logger.error(
            f"Empty quotes from Y-Finance [{ticker}]: range={range} | "
            f"interval={interval} | start={start} | end={end}"
        )
        raise EmptyResponse(f"Empty quotes: {response}")

    ohlc = result[0]["indicators"]["quote"][0]
    return timestamps, ohlc


@validate_arguments(config=dict(arbitrary_types_allowed=True))
async def get_bars(
    ticker: str,
    *,
    client: AsyncClient,
    start: Optional[datetime] = None,
    end: Optional[datetime] = None,
    range: Optional[ValidRange] = None,
    interval: ValidInterval = "1m",
    include_prepost: bool = False,
    coerce_interval: bool = True,
) -> Bars:

    assert (
        range is not None or start is not None
    ), "please provide either a range or start[-end] datetimes"

    if range is not None and coerce_interval:
        _valid_intervals = get_valid_intervals_for_range(interval, range)
    elif start is not None and coerce_interval:
        _valid_intervals = get_valid_intervals_for_datetime(interval, start)
    else:
        _valid_intervals = [interval]

    if interval not in _valid_intervals:
        logger.debug(
            f"[{ticker}]: {interval} interval is not supported by yahoo-finance "
            f"for provided range ({range}), or start/end datetimes ({start}/{end}). "
            f"Supported intervals for your request: {_valid_intervals}"
        )
    errors = []
    for _interval in _valid_intervals:
        try:
            timestamps, ohlc = await get_quotes(
                ticker,
                client=client,
                start=start,
                end=end,
                range=range,
                interval=_interval,
                include_prepost=include_prepost,
                coerce_interval=coerce_interval,
            )
            break
        except EmptyResponse as error:
            errors.append(error)
            continue
    else:
        raise EmptyResponse(errors)
    bars = Bars(
        __root__=[
            Bar(
                timestamp=timestamp,
                close=close if close is not None else np.nan,
                high=high if high is not None else np.nan,
                low=low if low is not None else np.nan,
                open=open if open is not None else np.nan,
                volume=volume if volume is not None else np.nan,
                interval=_interval,
            )
            for timestamp, volume, open, close, low, high in zip(
                timestamps,
                ohlc["volume"],
                ohlc["open"],
                ohlc["close"],
                ohlc["low"],
                ohlc["high"],
            )
        ]
    )
    return bars


@validate_arguments(config=dict(arbitrary_types_allowed=True))
async def get_latest_bar(
    ticker: str,
    *,
    client: AsyncClient,
    interval: ValidInterval = "1m",
    include_prepost: bool = True,
) -> Bar:

    if interval.endswith("m") or interval.endswith("h"):
        range = "1d"
    elif interval.endswith("d"):
        range = "1mo"
    else:
        range = "6mo"
    bars = await get_bars(
        ticker,
        client=client,
        include_prepost=include_prepost,
        range=range,  # type: ignore
    )
    return bars[-1]


@validate_arguments(config=dict(arbitrary_types_allowed=True))
async def get_bar_at_timestamp(
    ticker: str,
    *,
    client: AsyncClient,
    timestamp: datetime,
) -> Bar:
    assert timestamp.tzinfo == pytz.utc, (
        f"Only timezone aware datetimes with tz={pytz.UTC} are supported, "
        f"got: timestamp<{timestamp.tzinfo}>"
    )
    interval: ValidInterval = "1m"
    start = timestamp.replace(second=0)

    bars = await get_bars(ticker, client=client, start=start, interval=interval)
    return bars[0]


async def get_closest_price_at_timestamp(
    ticker: str,
    *,
    client: AsyncClient,
    timestamp: datetime,
) -> float:
    bar = await get_bar_at_timestamp(ticker, client=client, timestamp=timestamp)
    bar_open_date = bar.timestamp
    bar_close_date = bar_open_date + time_symbol_to_timedelta(bar.interval)
    open_delta = timestamp - bar_open_date
    close_delta = bar_close_date - timestamp
    if open_delta < close_delta:
        return bar.open
    return bar.close


if __name__ == "__main__":
    from finvestor.core import setup_logging
    from finvestor.etoro.utils import parse_etoro_datetime

    setup_logging()

    async def main():
        ticker = "FB"
        async with AsyncClient() as client:
            date = parse_etoro_datetime("28/09/2021 15:13:03")
            bars = await get_bar_at_timestamp(ticker, client=client, timestamp=date)
            print(bars)

    asyncio.run(main())
