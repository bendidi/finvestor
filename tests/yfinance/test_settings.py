import pytest

from finvestor.yfinance.settings import YFSettings


@pytest.mark.disable_auto_env_setup
def test_yahoo_finance_settings_defaults(snapshot):
    assert YFSettings() == snapshot


def test_yahoo_finance_settings_env(snapshot):
    assert YFSettings() == snapshot
