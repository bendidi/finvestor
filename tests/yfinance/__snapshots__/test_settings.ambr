# name: test_yahoo_finance_settings_defaults
  <class 'dict'> {
    'ISIN_SCRAPE_MAX_RESULTS': 25,
    'ISIN_SCRAPE_TIMEOUT': 5,
    'ISIN_SCRAPE_URL': AnyHttpUrl('https://markets.businessinsider.com/ajax/SearchController_Suggest', scheme='https', host='markets.businessinsider.com', host_type='domain', path='/ajax/SearchController_Suggest'),
    'YAHOO_FINANCE_BASE_URL': AnyHttpUrl('https://query2.finance.yahoo.com', scheme='https', host='query2.finance.yahoo.com', host_type='domain'),
    'YAHOO_FINANCE_SCRAPE_URL': AnyHttpUrl('https://finance.yahoo.com/quote', scheme='https', host='finance.yahoo.com/quote', host_type='domain'),
  }
---
# name: test_yahoo_finance_settings_env
  <class 'dict'> {
    'ISIN_SCRAPE_MAX_RESULTS': 25,
    'ISIN_SCRAPE_TIMEOUT': 5,
    'ISIN_SCRAPE_URL': AnyHttpUrl('https://markets.businessinsider.com/ajax/SearchController_Suggest', scheme='https', host='markets.businessinsider.com', host_type='domain', path='/ajax/SearchController_Suggest'),
    'YAHOO_FINANCE_BASE_URL': AnyHttpUrl('https://some-overriden-url', scheme='https', host='some-overriden-url', host_type='int_domain'),
    'YAHOO_FINANCE_SCRAPE_URL': AnyHttpUrl('https://some-other-overriden-url', scheme='https', host='some-other-overriden-url', host_type='int_domain'),
  }
---
